Hw3

best-time-is-now opened the pull request.

best-time-is-now committed:
Scenario changes

best-time-is-now committed:
Homework 3

best-time-is-now committed:
Changes to fileReader, productPage and buyProduct_test

best-time-is-now committed:
checkElementExists helper adding

adudyak commented on helpers/fileReader.js:
```diff
@@ -0,0 +1,18 @@
+        return string.split("\r\n").map(Number);
+    },
+
+    addRandomElement2Array(array) {
```
why do you need to randomly duplicate one of the elements?

adudyak commented on tests/buyProduct_test.js:
```diff
@@ -12,16 +19,22 @@ const USER = {
+
+Data(productIds).Scenario('buy product', async ({ I, productPage, cartPage, current }) => {
+    I.amOnPage('/index.php?route=product/product&product_id=' + current);
+    console.log(await productPage.selectColor());
```
remove

adudyak commented on tests/buyProduct_test.js:
```diff
@@ -12,16 +19,22 @@ const USER = {
+Data(productIds).Scenario('buy product', async ({ I, productPage, cartPage, current }) => {
+    I.amOnPage('/index.php?route=product/product&product_id=' + current);
+    console.log(await productPage.selectColor());
+    console.log(await productPage.selectSize());
```
remove

adudyak commented on tests/buyProduct_test.js:
```diff
@@ -12,16 +19,22 @@ const USER = {
+    console.log(await productPage.selectColor());
+    console.log(await productPage.selectSize());
+
     await I.emptyCart();
```
empty cart after login

adudyak commented on steps_file.js:
```diff
@@ -37,8 +38,18 @@ module.exports = function () {
 
-    clickCheckoutButton() {
-      this.click(checkoutButton);
+    async clickCheckoutButton() {
```
if product in not available, we shouldn't have any further steps. Currently in scenario you are trying to complete checkout form , even if product is not available.

adudyak replied:
this method causes an error
![image](https://github.com/best-time-is-now/js_ndubovyk/assets/19595492/9d987280-49fa-49e0-ac5f-5dc15f56475a)

best-time-is-now committed:
Corrections were made

best-time-is-now committed:
The proceedToCheckout method is moved to productPage

best-time-is-now committed:
One random element is chosen to run the Scenario

best-time-is-now committed:
The emptyCart method was moved

adudyak commented on helpers/fileReader.js:
```diff
@@ -0,0 +1,16 @@
+        return string.split("\r\n").map(Number);
+    },
+
+    getRandomElement2Array(array) {
```
`getRandomElementFromArray()`

adudyak commented on tests/buyProduct_test.js:
```diff
@@ -12,21 +18,26 @@ const USER = {
-    cartPage.clickConfirmOrder();
-    cartPage.verifySuccessfulPurchase();
+    productPage.proceedToCheckout();
+    const isAccessible = await I.clickCheckoutButton();
```
`clickCheckoutButton()` should not return anything, that is click.
create separate `checkProductIsAvailable()`

adudyak commented on steps_file.js:
```diff
@@ -23,22 +24,29 @@ module.exports = function () {
-    clickCheckoutButton() {
-      this.click(checkoutButton);
+        if (notAvailabilityIndicatorIsVisible) {
+          throw new Error('Sorry, the product is not available');
```
why do you need to throw an error to catch it in same function?
`if (await this.checkElementExists(notAvailableProduct)){
console.log('Please, choose the available product');
return false;
}
return true;`
or if remove console.log
`return await this.checkElementExists(notAvailableProduct)`

adudyak commented on tests/buyProduct_test.js:
```diff
@@ -12,21 +18,26 @@ const USER = {
+});
+
+Data([FileReader.getRandomElement2Array(productIds)]).Scenario('buy product', async ({ I, productPage, cartPage, current }) => {
     await I.emptyCart();
```
move this to `Before`

best-time-is-now committed:
Corrections were made

best-time-is-now committed:
checkElementExists_helper: the `tryElementExist` method was changed
productPage: the `selectColor`, `selectSize` and `getProductPrice` methods were changed

adudyak commented on pages/product.js:
```diff
@@ -1,31 +1,64 @@
 const { I } = inject();
+const checkElementExists = async function (element) { await tryTo(() => I.seeElement(element)); }
```
should be moved to helper, but I was not able to make it working there.
remove this one and use `tryElementExist` from helper to check availability

adudyak commented on tests/buyProduct_test.js:
```diff
@@ -12,21 +18,29 @@ const USER = {
+    
+    const isNotAccessible = await productPage.checkProductIsNotAvailable();
+    if (isNotAccessible) {
+        productPage.throwIfNotAvailable(isNotAccessible);
```
don't fail test, if product is n/a, just end test(skip further steps)

adudyak commented on pages/product.js:
```diff
@@ -1,31 +1,64 @@
+    await I.click(this.cartIcon);
+  },
+
+  async clickCheckoutButton() {
```
remove async/await

adudyak commented on pages/product.js:
```diff
@@ -1,31 +1,64 @@
   },
-}
+
+  async proceedToCheckout() {
```
remove async/await

adudyak commented on pages/cart.js:
```diff
@@ -24,8 +24,7 @@ module.exports = {
   async fillCheckoutForm(customer) {
-    const number = await I.grabNumberOfVisibleElements(this.firstNameField);
-    if (number > 0) {
+    if (await I.checkElementExists(this.firstNameField)) {
```
use `I.tryElementExist()`

best-time-is-now committed:
productPage: the `proceedToCheckout` and `clickCheckoutButton` methods are changed, the `checkElementExists` method is removed
cartPage: the `fillCheckoutForm` method is changed

best-time-is-now committed:
productPage: the `checkProductIsAvailable` and `throwIfNotAvailable` methods are changed
buyProduct_test: the corrections were made

best-time-is-now committed:
The 'getTax' method is changed

best-time-is-now committed:
productPage: the `getTax` method is changed

best-time-is-now committed:
cartPage: 'total' locator is changed

best-time-is-now committed:
cartPage: 'successfulPurchaseText' locator is changed

best-time-is-now committed:
cartPage: `ecoTax` and `vat` locators are removed
